source/slang/slang-ir-autodiff-rev.cpp

```diff
@@ -0,0 +1,832 @@
+                                Diagnostics::internalCompilerError,
+                                "Unexpected instruction. Expected func or generic");
+                        }
+                    }
+                }
+            }
+            
+            auto followUpWorkList = _Move(backwardTranscriberStorage.followUpFunctionsToTranscribe);
+            for (auto task : followUpWorkList)
+            {
+                auto diffFunc = as<IRFunc>(task.differential);
+                SLANG_ASSERT(diffFunc);
+                auto primalFunc = as<IRFunc>(task.primal);
+                SLANG_ASSERT(primalFunc);
+
+                backwardTranscriberStorage.transcribeFunc(builder, primalFunc, diffFunc);
+            }
+
+            // Transcribing the function body really shouldn't produce more follow up function body work.
+            // However it may produce new `ForwardDifferentiate` instructions, which we collect and process
+            // in the next iteration.
+            SLANG_RELEASE_ASSERT(backwardTranscriberStorage.followUpFunctionsToTranscribe.getCount() == 0);
+
+        }
+        return true;
+    }
+
+    // Checks decorators to see if the function should
+    // be differentiated (kIROp_ForwardDifferentiableDecoration)
+    // 
+    bool isMarkedForBackwardDifferentiation(IRInst* callable)
+    {
+        return callable->findDecoration<IRBackwardDifferentiableDecoration>() != nullptr;
+    }
+
+    IRStringLit* getBackwardDerivativeFuncName(IRInst* func)
+    {
+        IRBuilder builder(&sharedBuilderStorage);
+        builder.setInsertBefore(func);
+
+        IRStringLit* name = nullptr;
+        if (auto linkageDecoration = func->findDecoration<IRLinkageDecoration>())
+        {
+            name = builder.getStringValue((String(linkageDecoration->getMangledName()) + "_bwd_diff").getUnownedSlice());
+        }
+        else if (auto namehintDecoration = func->findDecoration<IRNameHintDecoration>())
+        {
+            name = builder.getStringValue((String(namehintDecoration->getName()) + "_bwd_diff").getUnownedSlice());
+        }
+
+        return name;
+    }
+
+    ReverseDerivativePass(AutoDiffSharedContext* context, DiagnosticSink* sink) :
+        InstPassBase(context->moduleInst->getModule()),
+        sink(sink),
+        backwardTranscriberStorage(context, context->sharedBuilder, sink),
+        autodiffContext(context),
+        pairBuilderStorage(context)
+    {
+        backwardTranscriberStorage.pairBuilder = &pairBuilderStorage;
+    }
+
+protected:
+    // A transcriber object that handles the main job of 
+    // processing instructions while maintaining state.
+    //
+    BackwardDiffTranscriber         backwardTranscriberStorage;
+
+    // Diagnostic object from the compile request for
+    // error messages.
+    DiagnosticSink*                 sink;
+
+    // Builder for dealing with differential pair types.
+    DifferentialPairTypeBuilder     pairBuilderStorage;
```

Comment:
Should the diff pair lowering step be shared by both forward and reverse pass? It doesn't seem to me that this pass should own diff pair lowering.